### DISCUSSION TITLE
Cannot set __hash__ on my Base using dataclasses

### DISCUSSION THREAD

**Discussion by bmakan** [sqlalchemy/sqlalchemy#9503] [Usage Questions]
In some cases, I like to add my instances into a set, and for that I need `__hash__` to be implemented. I've added it to my `Base` class, but it's not there in the children's classes.

I understand dataclasses only add `__hash__` if it's immutable, but why is it removing my own `__hash__`? But it does work if I add `__hash__` directly to the model class.

```python
class Base(MappedAsDataclass, DeclarativeBase, init=False, kw_only=True):
    def __eq__(self, other):
        if self.__class__ != other.__class__:
            return False

        return hash(self) == hash(other)

    def __hash__(self):
        columns = get_columns(self)
        return hash(','.join([str(getattr(self, x.key)) for x in columns if x.primary_key]))

class Child(Base):
    ...

Child.__hash__ # None
```

**Comment by zzzeek** [comment]
this is the behavior of dataclasses, becasue you did not set up unsafe_hash=True

here's a pure DC version


```py
import dataclasses


@dataclasses.dataclass(init=False, kw_only=True)
class Base:
    def __eq__(self, other):
        if self.__class__ != other.__class__:
            return False

        return hash(self) == hash(other)

    def __hash__(self):
        return id(self)


@dataclasses.dataclass(init=False, kw_only=True)
class Child(Base):
    ...


print(Child.__hash__)  # None

```

would need to add 

```
@dataclasses.dataclass(init=False, kw_only=True, unsafe_hash=True)
class Child(Base):
    ...
```

**Reply by bmakan** [reply]
I tried that too, but then when I added my `__hash__`, it raised an error stating it cannot overwrite it.

In the end, I just dropped dataclasses completely. There seems to be too many limitations around them and I've spent too much time resolving those during this upgrade process (which would otherwise be quite painless considering you guys documented it very well!).

**Reply by zzzeek** [reply]
> I tried that too, but then when I added my `__hash__`, it raised an error stating it cannot overwrite it.

to where, to Child?   you can place a hash method as is....

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(MappedAsDataclass, DeclarativeBase, init=False, kw_only=True):
    def __eq__(self, other):
        if self.__class__ != other.__class__:
            return False

        return hash(self) == hash(other)

    def __hash__(self):
        return id(self)

class Child(Base):
    __tablename__ = 't'
    id: Mapped[int] = mapped_column(primary_key=True)

    def __hash__(self):
        return id(self)

print(Child.__hash__)
```

**Reply by bmakan** [reply]
I put it into the `Base` with the `unsafe_hash=True`
```python
class Base(MappedAsDataclass, DeclarativeBase, init=False, kw_only=True, unsafe_hash=True):
   ...
```

and got the following error:
```python
TypeError: Cannot overwrite attribute __hash__ in class Base
```

**Reply by zzzeek** [reply]
yeah this is all dataclasses, it looks like the rules for when it raises or not are determined by this simple lookup table:

https://github.com/python/cpython/blob/65fb7c4055f280caaa970939d16dd947e6df8a8d/Lib/dataclasses.py#L121-L154


i would study that to see exactly what the rules are

**Reply by bmakan** [reply]
Well, these are the cases where it doesn't raise an error, and the `frozen` is `False` (`frozen` is not supported by the SQLA and I don't want my objects to be read-only).
The default case should simply inherit from the base class (`object` most likely). But it says "not hashable" so I expect they do some voodoo to enforce it.

I also tried setting `eq=False` and provide my own `__eq__` to the `Base` class, but then a whole bunch of errors popped across my codebase.

```
# __hash__

#    +------------------- unsafe_hash= parameter
#    |       +----------- eq= parameter
#    |       |       +--- frozen= parameter
#    |       |       |
#    v       v       v    |        |        |
#                         |   no   |  yes   |  <--- class has explicitly defined __hash__
# +=======+=======+=======+========+========+
# | False | False | False |        |        | No __eq__, use the base class __hash__
# +-------+-------+-------+--------+--------+
# | False | True  | False | None   |        | <-- the default, not hashable
# +-------+-------+-------+--------+--------+
# For boxes that are blank, __hash__ is untouched and therefore
# inherited from the base class.  If the base is object, then
# id-based hashing is used.
```

Looks like I won't be using dataclasses if I want to be able to use `set`s with my instances.
